core/templates/dev/head/galleries/Gallery.js

```diff
@@ -203,6 +203,125 @@ oppia.controller('Gallery', [
+        $scope.tileDisplayCount = MAXTILEDISPLAY;
+      }
+      $('.oppia-gallery-tiles-carousel').css({
+        width: $scope.galleryTilesResizedWidth + 'px'
+      });
+
+      // The following determines whether to enable left scroll after resize
+      for (i = 0; i < $('.oppia-gallery-tiles-carousel').length; i++) {
+        var carouselJQuerySelect = '.oppia-gallery-tiles:eq(n)'.replace(
+          'n', i);
+        var carouselScroll = $(carouselJQuerySelect).scrollLeft();
+        var index = Math.ceil(carouselScroll / TILEDISPLAYWIDTH);
+        $scope.rowIndices[i] = index;
+      }
+    };
+
+    $scope.scroll = function(ind, isLeftScroll) {
+      var carouselJQuerySelect = '.oppia-gallery-tiles:eq(n)'.replace('n', ind);
+      var direction = isLeftScroll ? -1 : 1;
+      var currentScrollPosition = $(carouselJQuerySelect).scrollLeft();
+
+      // Scrolling eligibility checks
+      if ($scope.galleryGroups[ind].activity_summary_dicts.length <=
+          $scope.tileDisplayCount) {
+        return;
+      } else if (isLeftScroll && $scope.rowIndices[ind] == 0) {
+        return;
+      } else if (!isLeftScroll && $scope.rowIndices[ind] >=
+                 $scope.galleryGroups[ind].activity_summary_dicts.length -
+                 $scope.tileDisplayCount) {
+        return;
+      }
+
+      if (currentScrollPosition < 0) {
+        currentScrollPosition = 0;
+      }
+
+      if (isLeftScroll) {
+        $scope.rowIndices[ind] -= $scope.tileDisplayCount;
+        if ($scope.rowIndices[ind] < 0) {
+          $scope.rowIndices[ind] = 0;
+        }
+      } else {
+        $scope.rowIndices[ind] += $scope.tileDisplayCount;
+        if ($scope.rowIndices[ind] >
+            $scope.galleryGroups[ind].activity_summary_dicts.length - 1) {
+          $scope.rowIndices[ind] =
+            $scope.galleryGroups[ind].activity_summary_dicts.length -
+            $scope.tileDisplayCount + 1;
+        }
+      }
+      var newScrollPosition = currentScrollPosition +
+        ($scope.galleryTilesResizedWidth * direction);
+
+      var lftOverlaySelect = '.oppia-gallery-tiles-carousel-overlay-left:eq(n)'.
+        replace('n', ind);
+      var rtOverlaySelect = '.oppia-gallery-tiles-carousel-overlay-right:eq(n)'.
+        replace('n', ind);
+
+      $(carouselJQuerySelect).animate({
+        scrollLeft: newScrollPosition
+      }, {
+        duration: 800,
+        queue: false
+      });
+
+      $(lftOverlaySelect).css({
+        display: 'inline'
+      }).fadeOut({
+        duration: 800,
+        queue: false
+      });
+      $(rtOverlaySelect).css({
+        display: 'inline'
+      }).fadeOut({
+        duration: 800,
+        queue: false
+      });
+    };
+
+    // Keeps track of the left most indices of each group
+    $scope.rowIndices = {};
+    // Keeps track if site is loading for the first time.
+    var firstLoad = true;
+
+    $scope.$watch('galleryGroups', function() {
```

Comment:
Why would galleryGroups change? I think it's better to have an explicit callback once galleryGroups is loaded for the first time, i.e. in the callback for $http.get().